api-examples/transformer_utils.py

```diff
@@ -0,0 +1,422 @@
+    """Read each word, and produce a tensor of x and y that are identical
+    """
+    def __init__(self, nctx: int, use_subword: bool = False, model_file: str = None, vocab_file: str = None):
+        """Create a reader with a context window that reads words
+
+        :param nctx: The context window length
+        :param use_subword: If true, use BPE, else words
+        """
+        self.use_subword = use_subword
+        if self.use_subword:
+            vectorizer = BPEVectorizer1D(model_file=model_file, vocab_file=vocab_file)
+        else:
+            vectorizer = Token1DVectorizer(transform_fn=baseline.lowercase)
+        super().__init__(nctx, {'x': vectorizer})
+
+    def build_vocab(self, files):
+        """Read the vocab file to get the tokens
+
+        :param files:
+        :return:
+        """
+        if self.use_subword:
+            super().build_vocab(files)
+            return {'x': self.vectorizers['x'].vocab}
+        return super().build_vocab(files)
+
+    def load(self, filename, vocabs):
+        features = self.load_features(filename, vocabs)
+        x_tensor = torch.tensor(features['x'], dtype=torch.long)
+        batch_width = self.nctx * 2
+        num_sequences_word = (x_tensor.size(0) // batch_width) * batch_width
+        x_tensor = x_tensor.narrow(0, 0, num_sequences_word).view(-1, batch_width)
+        # Take the first half for x_tensor, and the second half for y_tensor
+        return TensorDataset(x_tensor[:, :self.nctx], x_tensor[:, self.nctx:])
+
+
+class MultiFileLoader(IterableDataset):
+
+    def __init__(self, directory, pattern, vocabs, vectorizer, nctx):
+        super().__init__()
+        self.vectorizer = vectorizer
+        self.pattern = pattern
+        self.nctx = nctx
+        self.directory = directory
+        self.vocab = vocabs
+        self.samples = 0
+        self.rank = 0
+        self.world_size = 1
+        if torch.distributed.is_initialized():
+            self.rank = torch.distributed.get_rank()
+            self.world_size = torch.distributed.get_world_size()
+
+        if os.path.exists(f"{directory}/md.yml"):
+            f = read_yaml(f"{directory}/md.yml")
+            self.samples = f['num_samples']
+        else:
+            files = list(glob.glob(f"{directory}/{self.pattern}"))
+            pg = create_progress_bar(len(files))
+            for file in pg(files):
+                with open(file) as rf:
+                    for _ in rf:
+                        self.samples += 1
+            write_yaml({'num_samples': self.samples}, f"{directory}/md.yml")
+
+    def __len__(self):
+        return self.samples
+
+    def __iter__(self):
+        # Each node has the same worker_info, so the unique offsets for each is
+        # rank * num_workers + worker_id
+        # and the total available workers is world_size * num_workers
+        worker_info = torch.utils.data.get_worker_info()
+        files = list(glob.glob(f"{self.directory}/{self.pattern}"))
```

Comment:
Does `glob.glob` return a deterministic ordering? If it doesn't (for example last mod time could change where a file turns up in the list) we could have two workers who read the files in different orders which could result in reading the same files with different workers and probably missing some other files.

It seems like it returns the same order of `ls -U` but I don't know if that changes. It might be safer (but a little slow)  to sort this list